bin/cloud_builder.sh

```diff
@@ -0,0 +1,152 @@
+# and proxy at:
+# src/proxy
+PROXY_PATH="../../../../src/proxy"
+TAG_NAME="0.0.0"
+BUILD_DEBIAN="true"
+BUILD_DOCKER="true"
+
+function usage() {
+  echo "$0
+    -b        opts out of building debian artifacts
+    -c        opts out of building docker artifacts
+    -o        path to store build artifacts
+    -p        path to proxy repo (relative to istio repo, defaults to ${PROXY_PATH} ) 
+    -t <tag>  tag to use (optional, defaults to ${TAG_NAME} )"
+  exit 1
+}
+
+while getopts bco:p:t: arg ; do
+  case "${arg}" in
+    b) BUILD_DEBIAN="false";;
+    c) BUILD_DOCKER="false";;
+    o) OUTPUT_PATH="${OPTARG}";;
+    p) PROXY_PATH="${PROXY_PATH}";;
+    t) TAG_NAME="${OPTARG}";;
+    *) usage;;
+  esac
+done
+
+[[ -z "${OUTPUT_PATH}" ]] && usage
+[[ -z "${PROXY_PATH}" ]] && usage
+
+# switch to the root of the istio repo
+ROOT="$(cd "$(dirname "${BASH_SOURCE[0]}")/.." && pwd)"
+cd $ROOT
+
+export GOPATH="$(cd "$ROOT/../../.." && pwd)":${ROOT}/vendor
+echo gopath is $GOPATH
+
+if [ ! -d "${PROXY_PATH}" ]; then
+  echo "proxy dir not detected at ${PROXY_PATH}"
+  usage
+fi
+
+# Proxy has some specific requirements for Bazel's
+# config (plus it's nicely places bazel in batch
+# mode) so this component gets built first.
+
+pushd "${PROXY_PATH}"
+
+# Use this file for Cloud Builder specific settings.
+# This file sets RAM sizes and also specifies batch
+# mode that should shutdown bazel after each call.
+echo 'Setting bazel.rc'
+cp tools/bazel.rc.cloudbuilder "${HOME}/.bazelrc"
+if [ "${BUILD_DEBIAN}" == "true" ]; then
+  ./script/push-debian.sh -c opt -v "${TAG_NAME}" -o "${OUTPUT_PATH}"
+fi
+popd
+
+# Pilot likes to check if the source tree is 'clean'
+# when it queries for version/source informatin.  Some
+# other components like littering the tree so it's better
+# to build pilot sooner than later.
+
+# Pilot build expects this file to exist.  The usual
+# approach of adding a symlink to the user's config file
+# doesn't help when the user doesn't have one.
+touch pilot/platform/kube/config
+
+# building //... results in dirtied files:
+# broker/pkg/model/config/mock_store.go
+# broker/pkg/platform/kube/crd/types.go
+# mixer/template/apikey/go_default_library_handler.gen.go
+# mixer/template/apikey/go_default_library_tmpl.pb.go
+# mixer/template/template.gen.go
+bazel build //pilot/...
+
+# bazel_to_go likes to run from dir with WORKSPACE file
+./bin/bazel_to_go.py
+# Remove doubly-vendorized k8s dependencies that confuse go
+rm -rf vendor/k8s.io/*/vendor
+
+# bazel_to_go.py dirties generated_files and lintconfig.json
+# it's easier to ask git to restore files than add
+# an option to bazel_to_go to not touch them
+git checkout generated_files
+git checkout lintconfig.json
+
+pushd pilot
+./bin/upload-istioctl -r -o "${OUTPUT_PATH}"
+# An empty hub skips the tag and push steps.  -h "" provokes unset var error msg so using " "
+if [ "${BUILD_DOCKER}" == "true" ]; then
+  ./bin/push-docker -h " " -t "${TAG_NAME}" -b -o "${OUTPUT_PATH}"
```

Comment:
Do you know why we still have script in each subdir instead of having one common for the repo ?

Reply:
The push-docker scripts used to be standalone but in monorepo they do tend to share a common istio/bin/docker_lib.sh (the debian scripts are still separate and standalone but they basically just run "bazel build").  The remainder in each of the push-docker files is mostly istio-component-specific about which components to build (and how/where to build them).  A product build would benefit from a single file, but devs probably like being able to build just the stuff that applies to their component.  The main alternatives I can think of is a common script that takes an optional parameter on what to build, or place targets in the top-level Makefile (TBD on feasibility on foregoing scripts entirely), but either way that seems like a separate feature/enhancement task.